test/libsolidity/Metadata.cpp

```diff
@@ -458,34 +458,73 @@ BOOST_AUTO_TEST_CASE(metadata_optimiser_sequence)
+		settingsToTest.emplace_back(OptimiserSettings::standard());
+		settingsToTest.back().yulOptimiserCleanupSteps = "D";
+	}
+
+	{
+		// test that a custom optimizer sequence does not lead to default standard optimiser settings in metadata
+		settingsToTest.emplace_back(OptimiserSettings::standard());
+		settingsToTest.back().yulOptimiserSteps = "dhfoDgvulfnTUtnIf";
+	}
+
+	auto generateMetadataJson = [](std::string const& _source, OptimiserSettings const& _optimizerSettings) -> Json
 	{
-		OptimiserSettings optimizerSettings = OptimiserSettings::minimal();
-		optimizerSettings.runYulOptimiser = true;
-		optimizerSettings.yulOptimiserSteps = _optimizerSequence;
-		optimizerSettings.yulOptimiserCleanupSteps = _optimizerCleanupSequence;
 		CompilerStack compilerStack;
-		compilerStack.setSources({{"", sourceCode}});
+		compilerStack.setSources({{"", _source}});
 		compilerStack.setEVMVersion(solidity::test::CommonOptions::get().evmVersion());
-		compilerStack.setOptimiserSettings(optimizerSettings);
+		compilerStack.setOptimiserSettings(_optimizerSettings);
 
 		BOOST_REQUIRE_MESSAGE(compilerStack.compile(), "Compiling contract failed");
 
 		std::string const& serialisedMetadata = compilerStack.metadata("C");
 		Json metadata;
 		BOOST_REQUIRE(util::jsonParseStrict(serialisedMetadata, metadata));
 		BOOST_CHECK(solidity::test::isValidMetadata(metadata));
-		BOOST_CHECK(metadata["settings"]["optimizer"].contains("details"));
-		BOOST_CHECK(metadata["settings"]["optimizer"]["details"].contains("yulDetails"));
-		BOOST_CHECK(metadata["settings"]["optimizer"]["details"]["yulDetails"].contains("optimizerSteps"));
+		return metadata;
+	};
 
-		std::string const metadataOptimizerSteps = metadata["settings"]["optimizer"]["details"]["yulDetails"]["optimizerSteps"].get<std::string>();
-		std::string const expectedMetadataOptimiserSteps = _optimizerSequence + ":" + _optimizerCleanupSequence;
+	auto checkCustomSettings = [generateMetadataJson, sourceCode](OptimiserSettings const& _optimizerSettings)
+	{
+		auto const metadataJson = generateMetadataJson(sourceCode, _optimizerSettings);
+		BOOST_CHECK(metadataJson["settings"]["optimizer"].contains("details"));
+		BOOST_CHECK(metadataJson["settings"]["optimizer"]["details"].contains("yulDetails"));
+		BOOST_CHECK(metadataJson["settings"]["optimizer"]["details"]["yulDetails"].contains("optimizerSteps"));
+
+		std::string const metadataOptimizerSteps = metadataJson["settings"]["optimizer"]["details"]["yulDetails"]["optimizerSteps"].get<std::string>();
+		std::string const expectedMetadataOptimiserSteps = _optimizerSettings.yulOptimiserSteps + ":" + _optimizerSettings.yulOptimiserCleanupSteps;
 		BOOST_CHECK_EQUAL(metadataOptimizerSteps, expectedMetadataOptimiserSteps);
 	};
 
-	for (auto const& [sequence, cleanupSequence] : sequences)
-		check(sequence, cleanupSequence);
+	for (auto const& optimizerSettings: settingsToTest)
+		checkCustomSettings(optimizerSettings);
```

Comment:
Just a small nitpick about writing Boost tests. I never liked these black-box-style check/validation helpers. It's much better to have helpers that build values and leave the comparison to Boost macros. If it fails, you get a line number of the failing invocation of the macro, rather than helper definition, which is useless.

```suggestion
		BOOST_CHECK(metadataOptimizerSteps(settingsToTest) == expectedMetadataOptimiserSteps(settingsToTest));
```

Though in this case it wouldn't be enough because we also have a loop. It would have to be unrolled for best effect. For this reason I guess we can keep it as is, though if I were rewriting this, I'd have used functional-style helpers with no side-effects. It just seems like an obviously superior design.

Reply:
Another advantage of this is that such helpers are usually more generic and it's easier to reuse them in different tests. And having the macros in shared helpers would be especially annoying because then the failure does not even point you at the test.

This would allow you to put the loop below in a separate test case (which IMO it should be, it's completely independent from the other check, other than using the same helper).

Reply:
Fully agreed. I had initially also thought about this but decided against it because, well, larger diff and all.